File: backend/app/schemas/credits.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CreditOfferResponse(BaseModel):
    id: str
    sku: str = ""
    price_krw: int
    base_credits: int
    product_bonus_credits: int
    first_purchase_bonus_percent: int
    total_credits: int
    first_purchase_total_credits: int
    label: str
    payment_available: bool = False


class CreditFlowResponse(BaseModel):
    code: str
    label: str
    credits: int
    energy_percent: int
    energy_eligible: bool
    bonus_eligible: bool


class CreditCatalogResponse(BaseModel):
    credit_policy_version: str
    energy_policy_version: str
    offers: list[CreditOfferResponse]
    flows: list[CreditFlowResponse]


class CreditRewardMissionResponse(BaseModel):
    code: str
    credits: int
    completed: bool


class CreditBalanceResponse(BaseModel):
    purchased_credits: int
    bonus_credits: int
    debt_credits: int = 0
    total_credits: int
    energy_percent: int
    energy_max_percent: int
    next_energy_recovery_at: datetime | None
    credit_policy_version: str
    energy_policy_version: str
    reward_missions: list[CreditRewardMissionResponse] = Field(default_factory=list)


class CreditUsageResponse(BaseModel):
    id: str
    flow: str
    credits: int
    energy_percent: int
    bonus_credits: int
    purchased_credits: int
    status: str
    created_at: datetime


class CreditUsageListResponse(BaseModel):
    items: list[CreditUsageResponse] = Field(default_factory=list)


class CreditPurchaseGrantRequest(BaseModel):
    order_id: str = Field(min_length=1, max_length=80)


class CreditPurchaseGrantResponse(BaseModel):
    order_id: str
    status: str
    granted_credits: int
    purchased_credits: int
    bonus_credits: int
    debt_credits: int
    total_credits: int


class CreditPurchaseHistoryItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    provider_order_id: str
    sku: str
    status: Literal["processing", "granted", "refunded", "failed", "review"]
    base_credits: int
    product_bonus_credits: int
    first_purchase_bonus_credits: int
    granted_credits: int
    chargeback_credits: int
    created_at: datetime
    granted_at: datetime | None
    refunded_at: datetime | None


class CreditPurchaseHistoryResponse(BaseModel):
    items: list[CreditPurchaseHistoryItemResponse] = Field(default_factory=list)


class CreditPurchaseOperationsDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    provider_order_id: str
    user_id: UUID | None
    sku: str
    status: str
    provider_status: str
    price_krw: int
    base_credits: int
    product_bonus_credits: int
    first_purchase_bonus_credits: int
    granted_credits: int
    chargeback_credits: int
    failure_reason: str
    provider_checked_at: datetime | None
    granted_at: datetime | None
    refunded_at: datetime | None
    retention_until: datetime | None = None
    created_at: datetime | None = None


class CreditPurchaseOperationsAccount(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    purchased_credits: int
    bonus_credits: int
    debt_credits: int


class CreditPurchaseOperationsLedger(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    entry_type: str
    balance_type: str
    amount: int
    idempotency_key: str
    created_at: datetime


class CreditPurchaseOperationsResponse(BaseModel):
    purchase: CreditPurchaseOperationsDetail
    account: CreditPurchaseOperationsAccount | None
    ledger: list[CreditPurchaseOperationsLedger]


class CreditPurchaseOperationsQueueResponse(BaseModel):
    purchases: list[CreditPurchaseOperationsDetail]


class CreditPurchaseAuditItemResponse(BaseModel):
    purchase: CreditPurchaseOperationsDetail
    reasons: list[Literal["stale_processing", "status_review", "status_failed", "grant_amount_invalid", "purchase_ledger_mismatch", "refund_amount_mismatch", "chargeback_ledger_mismatch"]]


class CreditAccountAuditItemResponse(BaseModel):
    user_id: UUID
    purchased_credits: int
    bonus_credits: int
    debt_credits: int
    purchased_ledger_total: int
    bonus_ledger_total: int
    reasons: list[Literal["purchased_balance_mismatch", "bonus_balance_mismatch"]]


class CreditPurchaseAuditResponse(BaseModel):
    generated_at: datetime
    purchases: list[CreditPurchaseAuditItemResponse]
    accounts: list[CreditAccountAuditItemResponse]
    truncated: bool

File: apps/frontend/src/api/credits.ts
import { apiJson } from "./client.js";

export type CreditOffer = {
  id: string;
  sku: string;
  price_krw: number;
  base_credits: number;
  product_bonus_credits: number;
  first_purchase_bonus_percent: number;
  total_credits: number;
  first_purchase_total_credits: number;
  label: string;
  payment_available: boolean;
};

export type CreditFlow = {
  code: string;
  label: string;
  credits: number;
  energy_percent: number;
  energy_eligible: boolean;
  bonus_eligible: boolean;
  hard_daily_limit: number;
  intro_free_uses: number;
};

export type CreditCatalog = {
  credit_policy_version: string;
  energy_policy_version: string;
  offers: CreditOffer[];
  flows: CreditFlow[];
};

export type CreditUsage = {
  id: string;
  flow: string;
  credits: number;
  energy_percent: number;
  bonus_credits: number;
  purchased_credits: number;
  status: "reserved" | "committed" | "refunded";
  created_at: string;
};

export type CreditUsageList = { items: CreditUsage[] };

export const CREDIT_BALANCE_UPDATED_EVENT = "alive:credit-balance-updated";

export type CreditRewardMission = {
  code: "signup" | "first_character" | "first_dm";
  credits: number;
  completed: boolean;
};

export type CreditBalance = {
  purchased_credits: number;
  bonus_credits: number;
  debt_credits: number;
  total_credits: number;
  energy_percent: number;
  energy_max_percent: number;
  next_energy_recovery_at: string | null;
  credit_policy_version: string;
  energy_policy_version: string;
  reward_missions: CreditRewardMission[];
};

export function getCreditBalance(): Promise<CreditBalance> {
  return apiJson<CreditBalance>("/credits");
}

export function getCreditCatalog(): Promise<CreditCatalog> {
  return apiJson<CreditCatalog>("/credits/catalog");
}

export function getCreditUsage(): Promise<CreditUsageList> {
  return apiJson<CreditUsageList>("/credits/usage");
}

export type CreditPurchaseGrant = {
  order_id: string;
  status: "granted" | "refunded" | "processing" | "failed" | "review";
  granted_credits: number;
  purchased_credits: number;
  bonus_credits: number;
  debt_credits: number;
  total_credits: number;
};

export type CreditPurchase = {
  provider_order_id: string;
  sku: string;
  status: "processing" | "granted" | "refunded" | "failed" | "review";
  base_credits: number;
  product_bonus_credits: number;
  first_purchase_bonus_credits: number;
  granted_credits: number;
  chargeback_credits: number;
  created_at: string;
  granted_at: string | null;
  refunded_at: string | null;
};

export type CreditPurchaseList = { items: CreditPurchase[] };
export type TossIapEnvironment = "toss" | "sandbox";

export function grantCreditPurchase(orderId: string, sku: string = "", environment: TossIapEnvironment = "toss"): Promise<CreditPurchaseGrant> {
  return apiJson<CreditPurchaseGrant>("/credits/purchases/grant", { method: "POST", body: JSON.stringify({ order_id: orderId, sku, environment }) });
}

export function getCreditPurchases(): Promise<CreditPurchaseList> {
  return apiJson<CreditPurchaseList>("/credits/purchases");
}

export function notifyCreditBalanceUpdated(): void {
  if (typeof window === "undefined") return;
  window.dispatchEvent(new Event(CREDIT_BALANCE_UPDATED_EVENT));
}
